### ISSUE TITLE
`DOMDocument::prepend()` that's used in `image_add_svg_attributes()` does not work on PHP5 or PHP7

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#6577]
While working on #6487, the PR was failing with the following error:
```
???
```

Turns out that this is the problematic line in `core/includes/image.inc`:
```
$doc->firstChild->prepend($title);
```

You can test this by mimicking what `image_add_svg_attributes()` is trying to do:
1. Install the devel module on a Backdrop site that's running on PHP8.
2. Use the "Execute PHP code" feature that the devel module provides in `admin/devel/php`, and run the following code:
   ```php
   $svg_content = '<svg xmlns="http://www.w3.org/2000/svg" ><path d="blah"></path></svg>';

   $attributes = array(
     'alt' => "this is a title",
   );

   $doc = new DOMDocument();
   $doc->loadXML($svg_content);

   $title = $doc->createElement('title');
   $title->textContent = $attributes['alt'];
   $doc->firstChild->prepend($title);

   $result = $doc->saveXML($doc->firstChild);

   dpm($result);
   ```
   You should get the following result (the `<title>` HTML tag is inserted in the svg w/o any errors 👍🏼 ):
   `<svg xmlns="http://www.w3.org/2000/svg"><title>this is a title</title><path d="blah"/></svg>`
4. Repeat the same process on a site that's running PHP5 or PHP7 -> errors 👎🏼 

Checking https://www.php.net/manual/en/domdocument.prepend.php shows that `DOMDocument::prepend()` is only available in PHP8:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/86c6fe27-f484-4e3c-8a94-a4cd06283a5f)

We can use [`DOMNode::insertBefore()`](https://www.php.net/manual/en/domnode.insertbefore.php) instead, which is available for PHP5 through to v8 (and which I've tested in the PR for #6487 and all tests pass then).

**Comment by klonos** [comment]
...I'll try to add a failing test first, before fixing this.

**Comment by indigoxela** [comment]
@klonos many thanks for spotting this (severe) PHP compatibility problem.

I can see, this is yet another issue, you assigned yourself to. So I volunteer to take that one off your shoulders.
A [PR](https://github.com/backdrop/backdrop/pull/4771) is available for testing and review.

Functional tests extended, too, to check for icon contents after rendering with different option sets.

**Comment by klonos** [comment]
Thanks @indigoxela 🙏🏼 ...this will unblock my PR for #6487, which currently includes a temporary fix for this. I've left some suggestions to consider in the PR.

**Comment by indigoxela** [comment]
@klonos actually, taking another look... property firstChild seems OK, also with PHP 5.6. Not sure, why this didn't work initially. This allows to simplify the code a bit.
And, yes, a hint that we can further simplify it when we drop support for PHP 7.x makes sense.

**Comment by klonos** [comment]
> ...a hint that we can further simplify it when we drop support for PHP 7.x makes sense.

That's right. I was actually going for "don't change this - it will break things" for the inline comment, but your thinking is better. We should actually add a `@todo` that says something along the lines of `@todo: Change this to use DOMDocument::prepend() once we drop support for PHP5 and PHP7.`

**Comment by indigoxela** [comment]
> ``@todo:` Change this to use DOMDocument::prepend() once we drop support for PHP5 and PHP7.

Are you sure?
The code change and benefit would be minimal. Seems a bit odd to make this a TODO - as if this were broken with PHP 8+, which isn't the case.

**Comment by klonos** [comment]
> The code change and benefit would be minimal.

Fair enough. Also, why would we intentionally break things with existing sites when we have a way that works with multiple PHP versions.

**Comment by quicksketch** [comment]
Thanks @indigoxela and @klonos! This seems like a very important bug fix. Looks like the current test coverage failed to test the "alt" text capability. Thank you for adding the new test coverage.

**Comment by argiepiano** [comment]
The solution provided in this [PR](https://github.com/backdrop/backdrop/pull/4771) solves the immediate problem by using the correct method, but it assumes`firstChild` will always be the first element of the DOMDocument - this may or may not be true. The only way to assure we are dealing with the first element is by getting it with `firstElementChild`.

In addition that PR does not fix another problem: the code still tries to use methods `setAttribute` and `getAttribute` later in that function on a `DOMNode` object. According to the [manual](https://www.php.net/manual/en/book.dom.php), those methods do not exist for DOMNode, which may eventually cause issues if the first child is not the first element. Again, the way to assure we are using those methods correctly is by using `firstElementChild` instead of `firstChild`. I'll provide an alternative PR in the near future.

**Comment by klonos** [comment]
thanks for clarifying this @argiepiano 🙏🏼 ...just trying to get my head around this, is the difference that `firstChild` finds any child, even if not an element/tag (like plain text or inline comment), whereas `firstElementChild` makes sure that it selects a "tag"?

**Comment by klonos** [comment]
...right. There's good explanations here:
- https://www.w3schools.com/jsref/prop_node_firstchild.asp
  > ### `firstChild` vs `firstElementChild`
  > `firstChild` returns the first child `node` (an element node, a text node or a comment node). Whitespace between elements are also text nodes.
  >
  > `firstElementChild` returns the first child `element` (not text and comment nodes).
- https://stackoverflow.com/questions/51280481/what-is-the-difference-between-firstchild-and-firstelementchild
  > The difference boils down to the difference between a [Node](https://developer.mozilla.org/en/docs/Web/API/Node) and an [Element](https://developer.mozilla.org/en/docs/Web/API/Element). Every Element is also a Node, but a 
Node can also be a Text node, a Comment node, or something else.
  >
  > [firstElementChild](https://developer.mozilla.org/en/docs/Web/API/Element/firstElementChild) returns the first child _Element_ node, whereas [firstChild](https://developer.mozilla.org/en/docs/Web/API/Node/firstChild) returns the first child Node, regardless of whether it’s an Element, a Text, a Comment, or something else.

**Comment by indigoxela** [comment]
@argiepiano many thanks for joining discussion here.

> firstChild will always be the first element of the DOMDocument - this may or may not be true

You're right, of course, but I'm also asking myself, how tolerant we should be with icon file contents. Currently we're assuming, the first (probably only) item is the `<svg>` tag with its contents, which is true for all Phosphor icons in core, and is also true for all svg icons I ever created (that's a lot).
But we don't enforce anything with hook_icon_info(), so we better be safe than sorry, because the loop to add attributes is outside the try/catch.

**Comment by indigoxela** [comment]
Here we go, PR updated.

The most probable case IMO is some sort of "vendor" comment at the beginning of the file. The updated code makes sure, we act on the "svg" item, regardless of whether it's the first child or not. And also adds a test case for exactly that (a comment).

@argiepiano does this address your concerns? Or did you run into a special problem and want to share your use-case?

**Comment by argiepiano** [comment]
> Here we go, PR updated.

Looks great to me! Great thinking about using `$svg_tag = $doc->getElementsByTagName('svg')->item(0)` to get the actual svg element.

**Comment by argiepiano** [comment]
Question: should we also check to be sure `$svg_tag = $doc->getElementsByTagName('svg')->item(0)` actually returns something? For example, if the $svg_content actually doesn't have an svg tag. Or is that checked before calling this function?

**Comment by indigoxela** [comment]
>  should we also check to be sure $doc->getElementsByTagName('svg')->item(0) actually returns something?

@argiepiano  I scratched my head over that, but decided, that this function's not actually responsible to check, whether the svg file contains svg. If we decided to do such low-level checks on icon files, they should probably go to a validation step somewhere else, as this feels too late, when attempting to add/remove attributes.

**Comment by indigoxela** [comment]
@klonos I added another test to check a comment as first child, as requested.

**Comment by klonos** [comment]
Thanks @indigoxela 🙏🏼 ...this issue here is one of the blockers for #6487 that I've been working on. So I'll test and review soon.

**Comment by indigoxela** [comment]
> I'll test and review soon.

Yes, please! :pray: You know, time flies, and we really want to get this into 1.28.1.

**Comment by klonos** [comment]
@indigoxela the code looks good, and this is almost RTBC! ...I've applied the changes in `image.inc` from your PR to my PR for #6487, and all tests passed 👍🏼 ...however, PHPCS was being tripped by an inline comment. I've left a comment in your PR with details.

Once that's been fixed, this is RTCB.

**Comment by indigoxela** [comment]
>  ...however, PHPCS was being tripped by

...by unrelated code neither introduced nor touched by myself, and irrelevant for the fix we try to get in here.
I tried to explain the reason, why I avoid unrelated changes in a comment on the PR.

**Comment by argiepiano** [comment]
If this is being held by a PHPCS error unrelated to the PR here, let's open a new issue, and get this past the finish line, @klonos.

**Comment by klonos** [comment]
Several lines of code that this PR changes are part of wrongly indented code (which makes them wrongly indented as well). If we don't fix the indentation now, it will get flagged by PHPCS in other PRs that change code around it (which is what PHPCS does). With the same reasoning, it will be irrelevant to fix the indentation in those PRs (even more so, since they will not be touching the wrongly indented code at all). So we won't fix this wrong indentation ever(?).

Whatever. I don't agree, but I'm fine with this, because I want to see this issue fixed (it's currently blocking another issue I would like to work on).

If the @backdrop/core-committers that do the final review are fine with, It's OK I guess 🤷🏼

**Comment by indigoxela** [comment]
> Whatever. I don't agree...

Of course, you don't. :wink: 

Now to the core of my hesitation: instead of just fixing the indentation, I'd rather remove the try/catch. I don't get, why it's there, anyway. And especially, why it's wrapped around insertBefore, but not around setAttribute.
Honestly, I don't get the logic. Because if the insert fails, something's weird with the whole svg content, which will also cause setAttribute to fail.
And even if it were wrapped around the whole block (insertBefore, setAttribute), there would still be failure when attempting saveXML. And what's the expected outcome, anyway, if the "catch" strikes. But that would require more discussion, or that the reason for that specific try/catch location is explained.
That would take more time, though (which we don't have right now).

My suggestion would be a follow-up to not further block this bugfix. And fix the immediate problem ASAP.

**Comment by klonos** [comment]
> My suggestion would be a follow-up to not further block this bugfix. And fix the immediate problem ASAP.

That works for me 👍🏼

**Comment by herbdool** [comment]
RTBC

**Comment by quicksketch** [comment]
Thank you @indigoxela, @argiepiano, @klonos, and @herbdool! I appreciate the thorough review and testing. This looks great to me. I merged https://github.com/backdrop/backdrop/pull/4771 into 1.x and 1.28.x.

**Comment by indigoxela** [comment]
FTR: I created a follow-up for the try/catch (indent or logic): https://github.com/backdrop/backdrop-issues/issues/6613
